### PR TITLE
remove torch.equal usages

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89559
* __->__ #89527
* #89620
* #89526
* #89525

Preparation for the next PR in this stack: #89559.

I replaced

- `self.assertTrue(torch.equal(...))` with `self.assertEqual(..., rtol=0, atol=0, exact_device=True)`,
- the same for `self.assertFalse(...)` with `self.assertNotEqual(...)`, and
- `assert torch.equal(...)` with `torch.testing.assert_close(..., rtol=0, atol=0)` (note that we don't need to set `check_device=True` here since that is the default).

There were a few instances where the result of `torch.equal` is used directly. In that cases I've replaced with `(... == ...).all().item()` while sometimes also dropping the `.item()` depending on the context.

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

**Comment by kit1980** [comment]
@pmeier Do you want to eventually remove `torch.equal`? I see many Meta internal and open source projects outside PyTorch itself use it.

**Comment by pmeier** [comment]
@kit1980 Yes, that is the ultimate plan. The next PR in this stack will deprecate it with a detailed explanation on why we (`torch.testing` team) think this is reasonable. cc @mruberry

Given the widespread use that you mention, it is possible that the removal will never be acted on. The same happened to `torch.testing.assert_allclose` in #87974. We had an RFC (#61844) in which we and the community decided the change is worth it. We went through the whole deprecation cycle although we technically didn't need to. Still, the removal was reverted with very little chance of being re-instated. Thus, we are just going to document the function in #89526, but keep the deprecation. I'd wager a guess that something similar is going to happen for `torch.equal`.

I'm only going to send the next PR after I have battled and defeated CI here. If you want, I can ping you if that happens.

File: test/distributed/fsdp/test_fsdp_clip_grad_norm.py [hunk 1]
```diff
@@ -191,12 +191,12 @@ def _test_ddp_parity(
 
         # Check that the gradients were modified by `clip_grad_norm_()`
         for param, orig_grad in zip(ddp_model.parameters(), orig_ddp_grads):
-            assert not torch.equal(param.grad, orig_grad)
+            self.assertNotEqual(param.grad, orig_grad, rtol=0, atol=0, exact_device=True)
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_clip_grad_norm.py:194]: Thanks for the PR!

As the person who wrote this test, I wanted to mention that I had `assert not torch.equal(param.grad, orig_grad)` using `assert` and not `self.assertNotEqual` because the unit test would not be meaningful if the `assert` failed. (E.g. in this case, it would mean the gradients were not actually clipped because their norms were not sufficiently large.)

This is the general practice that the FSDP developers have been following: Use `assert` for internal unit test checks, where the `assert` should _never_ fail, and use `self.assertEqual` and variants for actual assertions representing the unit test.

I have no issue with the change in this PR, but I was curious if you had any thoughts on this matter.

Comment (pmeier) [reply test/distributed/fsdp/test_fsdp_clip_grad_norm.py:194]: Not sure I understand. What I got is that the plain `assert`'s should never fail since they are not directly related to the test at hand. Is that correct?

If yes, I think there is a better way solving this. The only way these plain asserts can trigger if something somewhat unrelated to the test broke. But this part should also be tested, right? So if something breaks there, we already have a test that signals this. Meaning, in this test here we can simply assume that everything unrelated is working correctly and just drop the assertion all together, no?

Comment (awgu) [reply test/distributed/fsdp/test_fsdp_clip_grad_norm.py:194]: To make this concrete, let me take this gradient norm clipping example, where we have a custom implementation to clip gradient norms but want it to match the local `nn.utils.clip_grad_norm_()`. (Ignore the details for why we want to do this.)

Suppose you wrote a unit test that runs forward/backward to generate gradients and then clipped them, for both the custom and baseline. If the generated gradients did not have sufficiently large norm, then the unit test would be vacuous since the gradient clipping was a no-op, and a passing unit test would not provide real signal.

The `assert`s were (1) to check that the gradients were indeed clipped, even for baseline, and (2) to signal to other developers reading the code that this was the case.

(1) is more of a sanity check for the unit test writer and by itself does not necessarily warrant keeping the `assert` in the unit test. However, I feel that (2) justifies keeping the `assert`. The `assert`s, despite being expected to not fail, can help inform readers of possibly non-obvious invariants or confirm that the test is working as expected.

The main reason I distinguish between `assert` and `self.assertEqual` variants is so that if somehow the `assert` fails (since something in the outside world changed), we know pretty fast that it is not due to the gradient norm clipping implementation (or whatever the unit test is testing) but rather something else. I leave the `self.assertEqual` variants for testing the gradient norm clipping implementation (or whatever the unit test is testing).

Does this make more sense?

Comment (pmeier) [reply test/distributed/fsdp/test_fsdp_clip_grad_norm.py:194]: I think I understand the use case now. Instead of this fine distinction, wouldn't it be better to

1. Use something else than `AssertionError` since they are somewhat "reserved" for the actual test failures. For example, `pytest` uses [`pytest.UsageError`](https://docs.pytest.org/en/7.2.x/reference/reference.html#pytest.UsageError) and `assert_close` / `assertEqual` use [`ValueError`](https://github.com/pytorch/pytorch/blob/860bae49e4925868a0221ec4345d08407280bac7/torch/testing/_comparison.py#L1199-L1200) to indicate something in the setup is wrong.
2. Add a descriptive error message what this error means.

Basically, replace `assert torch.equal(...)` with

```py
if not torch.eq(...).all():
    raise RuntimeError("The generated gradients did not have sufficiently large norm!")
```

Comment (awgu) [reply test/distributed/fsdp/test_fsdp_clip_grad_norm.py:194]: Great point! Your suggestion sounds good to me. I can follow up on these `assert` usages in the FSDP unit tests. Feel free to leave the change in this PR as is.

**Review by mruberry — APPROVED** [review]
Stamped!

**Comment by pmeier** [comment]
Test failures are unrelated:

- No GPU: https://github.com/pytorch/pytorch/actions/runs/3590839317/jobs/6045030583#step:10:30672
- Timeout: https://github.com/pytorch/pytorch/actions/runs/3590839317/jobs/6044676262#step:1:40
- Dependency issue: https://github.com/pytorch/pytorch/actions/runs/3590839327/jobs/6046456211#step:15:375

@pytorchbot merge -f "flaky CI"

**Comment by clee2000** [comment]
@pytorchbot revert -m "broke periodic multigpu tests  https://hud.pytorch.org/pytorch/pytorch/commit/4095ef8b809f922f2e0e09011afd00037d20a771
https://github.com/pytorch/pytorch/actions/runs/3592806602/jobs/6049368502" -c nosignal
